core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala

```diff
@@ -185,16 +186,34 @@ class PaigesScalacheckTest extends AnyFunSuite {
      * b.grouped + flatten(c) == (b + flatten(c)).grouped
      * flatten(c) + b.grouped == (flatten(c) + b).grouped
      */
-    forAll { (b: Doc, c: Doc) =>
+    def law(b: Doc, c: Doc): Assertion = {
       val flatC = c.flatten
-      val left = (b.grouped + flatC)
+
+      val left = b.grouped + flatC
       val right = (b + flatC).grouped
-      assert(left === right)
-      assert((flatC + b.grouped) === (flatC + b).grouped)
+      assert(left === right, debugEq(left, right))
+
+      if (!flatC.containsHardNewline) {
+        val lhs0 = flatC + b.grouped
+        val rhs0 = (flatC + b).grouped
+        assert(lhs0 === rhs0, debugEq(lhs0, rhs0))
+      }
+
       // since left == right, we could have used those instead of b:
-      assert((left.grouped + flatC) === (right + flatC).grouped)
+      val lhs1 = left.grouped + flatC
+      val rhs1 = (right + flatC).grouped
+      assert(lhs1 === rhs1, debugEq(lhs1, rhs1))
+    }
+
+    import Doc._
+    val b0 = Concat(Text("b"), Union(Text(" "), FlatAlt(Line, Text(" "))))
+    val c0 = Union(Concat(Text("b"), Text(" ")), Concat(Text("b"), FlatAlt(Line, Text(" "))))
+    law(b0, c0)
```

Comment:
We shouldn't commit this "raw AST" example -- I'd prefer to either delete it or use our combinators to build it up.